npm/podController.go

```diff
@@ -0,0 +1,627 @@
+		return key, needSync
+	}
+
+	needSync = true
+	return key, needSync
+}
+
+func (c *podController) addPod(obj interface{}) {
+	key, needSync := c.needSync("ADD", obj)
+	if !needSync {
+		log.Logf("[POD ADD EVENT] No need to sync this pod")
+		return
+	}
+	// K8s categorizes Succeeded and Failed pods as a terminated pod and will not restart them
+	// So NPM will ignorer adding these pods
+	podObj, _ := obj.(*corev1.Pod)
+	if podObj.Status.Phase == v1.PodSucceeded || podObj.Status.Phase == v1.PodFailed {
+		return
+	}
+	c.workqueue.Add(key)
+}
+
+func (c *podController) updatePod(old, new interface{}) {
+	oldPod := old.(*corev1.Pod)
+	newPod := new.(*corev1.Pod)
+
+	if oldPod.ResourceVersion == newPod.ResourceVersion {
+		// Periodic resync will send update events for all known pods.
+		// Two different versions of the same pods will always have different RVs.
+		log.Logf("[POD UPDATE EVENT] Two pods have the same RVs")
+		return
+	}
+
+	key, needSync := c.needSync("UPDATE", new)
+	if !needSync {
+		log.Logf("[POD UPDATE EVENT] No need to sync this pod")
+		return
+	}
+	c.workqueue.Add(key)
+}
+
+func (c *podController) deletePod(obj interface{}) {
+	podObj, ok := obj.(*corev1.Pod)
+	// DeleteFunc gets the final state of the resource (if it is known).
+	// Otherwise, it gets an object of type DeletedFinalStateUnknown.
+	// This can happen if the watch is closed and misses the delete event and
+	// the controller doesn't notice the deletion until the subsequent re-list
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			metrics.SendErrorLogAndMetric(util.NpmID, "[POD DELETE EVENT] Pod: Received unexpected object type: %v", obj)
+			return
+		}
+
+		if podObj, ok = tombstone.Obj.(*corev1.Pod); !ok {
+			metrics.SendErrorLogAndMetric(util.NpmID, "[POD DELETE EVENT] Pod: Received unexpected object type (error decoding object tombstone, invalid type): %v", obj)
+			return
+		}
+	}
+
+	log.Logf("[POD DELETE EVENT] for %s in %s", podObj.Name, podObj.Namespace)
+	if isHostNetworkPod(podObj) {
+		log.Logf("[POD DELETE EVENT] HostNetwork POD IGNORED: [%s/%s/%s/%+v%s]", podObj.UID, podObj.Namespace, podObj.Name, podObj.Labels, podObj.Status.PodIP)
+		return
+	}
+
+	var err error
+	var key string
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		utilruntime.HandleError(err)
+		metrics.SendErrorLogAndMetric(util.PodID, "[POD DELETE EVENT] Error: podKey is empty for %s pod in %s with UID %s",
+			podObj.ObjectMeta.Name, util.GetNSNameWithPrefix(podObj.Namespace), podObj.UID)
+		return
+	}
+
+	// (TODO): Reduce scope of lock later
+	c.npMgr.Lock()
+	defer c.npMgr.Unlock()
+
+	cachedNpmPodObj, npmPodExists := c.npMgr.PodMap[key]
+	if !npmPodExists {
+		return
+	}
+
+	// if the podIp exists, it must match the cachedIp
+	// (TODO): when does this situation happen? Is it ok to add it to workqueue?
+	if len(podObj.Status.PodIP) > 0 && cachedNpmPodObj.PodIP != podObj.Status.PodIP {
```

Comment:
we should add it to the WQ, but we will need to have a similar check, where we are deleting the cachedNpmPodobj and not the new PodIP.

this will happen if a Pod NSA/A is deleted but NPM missed the event, and then NSA/A is created with a new IP and is deleted again. This may not be an issue in WQ's scenario, but is a good check to have where we are cleaning up the cachedPod objs and not the state given to us